src/cabal

```diff
@@ -125,6 +127,53 @@ run_test () {
 }
 
 #
+# Cabal test for CI build bots.
+#
+run_testci () {
+    # Create a temporary directory which will be deleted when the script
+    # terminates. Unfortunately `mktemp` doesn't behave the same on
+    # Linux and OS/X so we need to try two different approaches.
+    CABAL_EXTRAS_TEMP=$(mktemp -d 2>/dev/null || mktemp -d -t 'cabal-extras')
+    trap "rm -rf $CABAL_EXTRAS_TEMP" EXIT
+
+    # Write out a haskell program which performs the control character
+    # processing.
+    UNCTL=$CABAL_EXTRAS_TEMP/unctl.hs
+    cat << EOF > $UNCTL
+import System.IO
+
+main :: IO ()
+main = do
+    hSetBuffering stdin  LineBuffering
+    hSetBuffering stdout LineBuffering
+    interact (go [])
+  where
+    go :: [Char] -- ^ current line
+       -> [Char] -- ^ stdin
+       -> [Char] -- ^ stdout
+
+    -- backspace - delete previous character
+    go (_ : line) ('\b' : xs) = go line xs
+    go []         ('\b' : xs) = go []   xs
+
+    -- carriage return - delete the whole line
+    go _          ('\r' : xs) = go []   xs
+
+    -- line feed - emit the current line
+    go line       ('\n' : xs) = reverse ('\n' : line) ++ go [] xs
+
+    -- normal character - add to current line
+    go line       (x    : xs) = go (x : line) xs
+
+    -- end of stream - emit the current line
+    go line       []          = line
+EOF
+
+    # Run the tests, but fix up their output before emitting it.
+    run_test "$@" | runghc $UNCTL
```

Comment:
Losing exit code? You can work around by funnelling output through a named pipe so there isn't actually a pipeline here.